### DISCUSSION TITLE
Developers - how to contribute

### DISCUSSION THREAD

**Discussion by ValueRaider** [ranaroussi/yfinance#1084] [Development]
# Branches

To support rapid development without breaking stable versions, this project uses a two-layer branch model:
![image](https://github.com/ranaroussi/yfinance/assets/96923577/5afe5e2b-a43c-4a64-a736-a9e57fb5fe70)
([inspiration](https://miro.medium.com/max/700/1*2YagIpX6LuauC3ASpwHekg.png))

- **dev** - new features & some bug-fixes merged here, tested together, conflicts fixed, etc. Once stable, can merge to ...
- **main**: stable branch. Where PIP releases are created

Most of the time you want to target the **dev** branch, but default is **main** so remember to switch to **dev**.

**Exception:** can straight merge into **main** if:
- yfinance is massively broken
- or part of yfinance is broken and the fix is simple and clearly doesn't affect anything else

# I'm a GitHub newbie, how do I contribute code?

1. Fork this project. If already forked, remember to `Sync fork`
2. Implement your change in your fork, ideally in a specific branch
3. Create a Pull Request, from your fork to this project. If addressing an Issue, link to it

### [How to download & run a GitHub version of yfinance](https://github.com/ranaroussi/yfinance/discussions/1080)

# Documentation website

The new docs website https://ranaroussi.github.io/yfinance/index.html is generated automatically from code.
When you fetch a branch or PR, probably it also has changes to docs - to generate and view this:

```
pip install Sphinx
sphinx-build -b html doc/source doc/_build/html
python -m http.server -d ./doc/_build/html
# open "localhost:8000" in browser
```

# Unit tests

Tests have been written using the built-in Python module `unittest`. Examples:

- Run all price tests: `python -m unittest tests.test_prices`

- Run sub-set of price tests: `python -m unittest tests.test_prices.TestPriceRepair`

- Run a specific test: `python -m unittest tests.test_prices.TestPriceRepair.test_ticker_missing`

- Run all tests: `python -m unittest discover -s tests`

# Git stuff

### commits

To keep the Git commit history and [network graph](https://github.com/ranaroussi/yfinance/network) compact please follow these two rules:
- For long commit messages use this: `git commit -m "short sentence summary" -m "full commit message"`
- `squash` tiny/negligible commits back with meaningful commits, or to combine successive related commits. [Guide](https://docs.gitlab.com/ee/topics/git/git_rebase.html#interactive-rebase) but basically it's:
```
git rebase -i HEAD~2
git push --force-with-lease origin <branch-name>
```

### rebase
You might be asked to move your branch from `main` to `dev`. Make sure you have pulled **all** relevant branches then run:
```
git checkout <your branch>
git rebase --onto dev main <branch-name>
git push --force-with-lease origin <branch-name>
```

**Comment by jcenmiami** [comment]
Just an FYI, download Pandas 2.1.0 and I started getting iloc warnings while using YFinance (utils.py). It's just a matter of adding iloc[x] to the code. I added it to the library and the warnings disappeared. I downgraded to 1.5.3 and all is good while the next version of YFinance is ready. Thanks for everything, love this library.

**Reply by ValueRaider** [reply]
Is fixed in `dev` branch #1672. Simply haven't got round to releasing.

**Comment by s-kust** [comment]
How to run tests to make sure I don't break anything?

**Reply by ValueRaider** [reply]
Only just seen this. I've added a new section.

**Reply by busterbeam** [reply]
and it needs updating, as the tests still don't work properly. fresh clone and even after installing the extra packages not stated in `requirements.txt` does it do something. but yea, not a great start for any dev wanting to help.

**Comment by DanielGNorris** [comment]
I have been using this library for about a month now for a coding project. But now the `yf.download` function always returns an empty data-frame. Could I possibly been IP blocked and is there anyway around it?

**Reply by ValueRaider** [reply]
this is wrong place

**Reply by DanielGNorris** [reply]
Where should I comment this then?

**Comment by ValueRaider** [comment]
superseded by https://github.com/ranaroussi/yfinance/blob/main/CONTRIBUTING.md
